kotlinx-coroutines-core/common/src/channels/Channels.common.kt

```diff
@@ -189,11 +200,68 @@ public suspend inline fun <E> ReceiveChannel<E>.consumeEach(action: (E) -> Unit)
  * ```
  */
 public suspend fun <E> ReceiveChannel<E>.toList(): List<E> = buildList {
-    consumeEach {
-        add(it)
-    }
+    consumeEach(::add)
 }
 
+/**
+ * [Consumes][consume] the elements of this channel into the provided mutable collection.
+ *
+ * This is a convenience function equivalent to calling [consumeAsFlow]
+ * followed by [kotlinx.coroutines.flow.toCollection].
+ * Please use [consumeAsFlow] directly in scenarios where elements should undergo additional transformations
+ * before being added to the resulting collection.
+ *
+ * [consumeTo] attempts to receive elements and put them into the collection until the channel is
+ * [closed][SendChannel.close].
+ *
+ * If the channel is [closed][SendChannel.close] with a cause, [consumeTo] will rethrow that cause.
+ * However, the elements already received up to that point will remain in the collection.
+ *
+ * Since this function is implemented using [consume], the function is _terminal_.
+ * In practice, this means that if adding new elements to the collection fails with an exception,
+ * that exception will be used for [cancelling][ReceiveChannel.cancel] the channel and rethrown.
+ *
+ * The intended use case for this function is collecting the remaining elements of a closed channel
```

Comment:
Not necessarily "the remaining elements of a **closed** channel".

It is definitely for use when we know that the channel **is or will be closed** at some point.

Reply:
Your sample is of this use-case, but there are others.

Reply:
What you're describing is not the intended use case. If you only know the channel will be closed at some point, then `consumeTo` uses an unbound amount of memory, which we try to avoid. A possible use case is sending a bounded number of elements (for example, we know there will be at most 100 elements before the channel gets closed). I'd argue that channels are not the right tool for the job in that case, you should instead build a list on the sender's side and publish it as a whole to the receiver.

Reply:
If you have a fairly closed scope for channels in mind, could you elaborate such use-cases in the kdoc? Currently it leaves me puzzled as to why this is the only case presented, even though I can see that I can use it in other ways.

Reply:
Yes, I also can't verify that. Somewhere, there may and probably do exist reasonable usages I haven't thought of. This section of the text is descriptive ("here's what this function was designed to do"), not prescriptive ("here's the singular scenario in which you're allowed to call this function"). The purpose of this section is to help people that are unsure if this is the function they want: if this description matches their scenario, they will know that yes, they have found the right one. If not, they may need to dive in deeper and/or keep searching.